lib/pushy_client/protocol_handler.rb

```diff
@@ -487,14 +487,24 @@ def self.send_signed_message(socket, method, client_private_key, session_key, me
              when :hmac_sha256
                make_header_hmac(message, session_key)
              end
-      begin
-        Timeout.timeout(10) do
-          socket.send_string(auth, ZMQ::SNDMORE)
-          socket.send_string(message)
-        end
-      rescue Timeout::Error
-        Chef::Log.info("[#{client.node_name}] ZMQ socket timed out. Triggering reconfigure")
-        # we don't immediately reconfigure because this is likely to amplify any stampedes. 
+      # https://github.com/chuckremes/ffi-rzmq/blob/master/lib/ffi-rzmq/socket.rb
+      # send_string
+      #
+      # +flags+ may be ZMQ::DONTWAIT and ZMQ::SNDMORE.
+      #
+      # Returns 0 when the message was successfully enqueued.
+      # Returns -1 under two conditions.
+      # 1. The message could not be enqueued
+      # 2. When +flags+ is set with ZMQ::DONTWAIT and the socket returned EAGAIN.
+      #
+      # With a -1 return code, the user must check ZMQ::Util.errno to determine the
+      # cause.
+
+      socket.send_string(auth, ZMQ::SNDMORE | ZMQ::DONTWAIT)
+      rc = socket.send_string(message, ZMQ::DONTWAIT)
```

Comment:
I think this is the right thing to do, but as I discussed with @jeremymv2, this changes a fundamental premise we made when writing the client, in that we trusted 0mq to deliver things eventually and didn't pay much attention to connection drops. 

On one hand this has had some bad effects, such as queuing up storms of heartbeats when the server goes away, and getting away from that is a good thing.

But we probably will need to keep an eye on possible impacts. One is that will probably increase the number of heartbeats lost (because we're not queuing them for retry), with knock on effects around the server sending aborts because it thinks the client is dead. That probably can be fixed with a server config change, requiring more dropped heartbeats before giving up on a client.

Another issue is around client state changes. If we change state (such as completing a running job), and we drop the packet with that info, the server's model of the client state will be out of sync with the actual. If the server detects that, it will send an abort to reset it to a known state, and we will lose the information about the job completing. If this starts to be a problem we will probably have to retry sending those state change packets.

Reply:
@markan 

I've run some tests accounting for 4 scenarios:

1. This patch applied, no simulated network outages, 30 minutes of continuous 2 minute ping loops
2. This patch applied, simulated network outages, 30 minutes of continuous 2 minute ping loops
3. Unpatched version 2.4.8, no simulated network outages, 30 minutes of continuous 2 minute ping loops
4. Unpatched version 2.4.8, simulated network outages, 30 minutes of continuous 2 minute ping loops

The number of `Received abort` messages for each scenario:

Windows
1. 1
2. 4
3. 1
4. 4

Linux
1. 1
2. 6
3. 1
4. 4

Comparing results of scenarios 1 against 3 and 2 against 4, the number of received abort messages did not significantly change - increasing by 2 on the Linux client.